libbpf-tools/schedsnoop.c

```diff
@@ -0,0 +1,309 @@
+	{ "pid", 'p', "PID", 0, "Process PID to trace"},
+	{ "syscall", 's', NULL, 0, "Trace SYSCALL Info"},
+	{},
+};
+
+static struct env {
+	int nr_cpus;
+	int trace_syscall;
+	int target;
+} env;
+
+int ti_map_fd;
+int si_map_fd;
+int start_map_fd;
+int end_map_fd;
+bool exiting = false;
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int pid;
+	switch (key) {
+	case 'p':
+		if(arg == NULL) {
+			fprintf(stderr, "PID is required\n");
+			argp_usage(state);
+		} else {
+			pid = strtol(arg, NULL, 10);
+			if (pid <= 0) {
+				fprintf(stderr, "Invalid PID: %s\n", arg);
+				argp_usage(state);
+			}
+			env.target = pid;
+		}
+		break;
+	case 's':
+		env.trace_syscall = 1;
+		break;
+	case ARGP_KEY_END:
+		if (!env.target) {
+			fprintf(stderr, "No target PID\n");
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static inline int time_to_str(u64 ns, char *buf, size_t len)
+{
+
+	if (ns > 10 * NS_IN_SEC)
+		snprintf(buf, len, "%llus", ns / NS_IN_SEC);
+	else if (ns > 10 * NS_IN_MS)
+		snprintf(buf, len, "%llums", ns / NS_IN_MS);
+	else if (ns > 10 * NS_IN_US)
+		snprintf(buf, len, "%lluus", ns / NS_IN_US);
+	else
+		snprintf(buf, len, "%lluns", ns);
+
+	return 0;
+}
+
+static inline void pr_ti(struct trace_info *ti, char *opt, char *delay)
+{
+	char buf[27];
+	snprintf(buf, sizeof(buf), "%lluus", ti->ts / NS_IN_US);
+	printf("%-27sCPU=%-7dPID=%-7dCOMM=%-20s%-37s%-17s\n",
+				buf, ti->cpu, ti->pid, ti->comm, opt,
+				delay ? delay : "");
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct trace_info *ti = data;
+	char d_str[16];
+	char comm_buf[2 * TASK_COMM_LEN];
+	char func[80];
+	struct si_key sik;
+	u64 siv;
+	static u64 w_start, p_start, last_time;
+
+
+	time_to_str(ti->ts - last_time, d_str, sizeof(d_str));
+
+	switch (ti->type) {
+	case TYPE_MIGRATE:
+		w_start = p_start = ti->ts;
+		pr_ti(ti, "MIGRATE", NULL);
+		break;
+	case TYPE_ENQUEUE:
+		w_start = p_start = ti->ts;
+		printf("----------------------------\n");
+		pr_ti(ti, "ENQUEUE", NULL);
+		break;
+	case TYPE_WAIT:
+		if (ti->pid == env.target) {
+			w_start = ti->ts;
+			pr_ti(ti, "WAIT AFTER EXECUTED", d_str);
+		} else {
+			time_to_str(ti->ts - p_start,
+					d_str, sizeof(d_str));
+			pr_ti(ti, "PREEMPTED", d_str);
+		}
+		break;
+	case TYPE_EXECUTE:
+		if (ti->pid == env.target) {
+			time_to_str(ti->ts - w_start,
+					d_str, sizeof(d_str));
+			pr_ti(ti, "EXECUTE AFTER WAITED", d_str);
+		} else {
+			p_start = ti->ts;
+			pr_ti(ti, "PREEMPT", NULL);
+		}
+		break;
+	case TYPE_DEQUEUE:
+		if (ti->pid == env.target)
+			pr_ti(ti, "DEQUEUE AFTER EXECUTED", d_str);
+		else {
+			time_to_str(ti->ts - p_start,
+					d_str, sizeof(d_str));
+			pr_ti(ti, "DEQUEUE AFTER PREEMPTED", d_str);
+		}
+		break;
+	case TYPE_SYSCALL_ENTER:
+		siv = ti->ts;
+		sik.cpu = ti->cpu;
+		sik.pid = ti->pid;
+		sik.syscall = ti->syscall;
+		bpf_map_update_elem(si_map_fd, &sik, &siv, BPF_ANY);

```

Comment:
why is this done from user-space and not from BPF? there might be significant delays and it's just less performant overall (and causes extra syscall for every other syscall?)